### PR TITLE
test: add deploy test

### PR DESCRIPTION
**Pull request by 0618**
## Problem

- add `e2e-flow` test, which is a copy of `test-e2e/create-amplify.test`
- add `deploy` test to `create-amplify` script
- [ ]  make the deploy test work

**Issue number, if available:**

## Changes

**Corresponding docs PR, if applicable:**

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: .github/workflows/poc-e2e-flow-test.yml [hunk 1]
```diff
@@ -61,7 +62,7 @@ jobs:
           node-version: ${{ matrix.node-version }}
       - name: Restore Build Cache
         uses: ./.github/actions/restore_build_cache
-      - name: Run E2E tests with ${{ matrix.pkg-manager }}
+      - name: Run E2E flow tests with ${{ matrix.pkg-manager }}
         shell: bash
         run: |
-          PACKAGE_MANAGER_EXECUTABLE=${{matrix.pkg-manager}} npm run test:dir packages/integration-tests/src/test-e2e/create_amplify.test.ts
+          PACKAGE_MANAGER_EXECUTABLE=${{matrix.pkg-manager}} npm run test:dir packages/integration-tests/src/e2e-flow/e2e_flow.test.ts
```
Comment (sobolk) [inline .github/workflows/poc-e2e-flow-test.yml:68]: If we're going for separate test file in separate dir (i.e. not included in e2e run)
then perhaps we could just use this one https://github.com/aws-amplify/amplify-backend/blob/main/packages/integration-tests/src/test-live-dependency-health-checks/health_checks.test.ts 
but adapt it to take package manager as input.

**Review by sobolk — COMMENTED** [review]
I do like an idea to just craft separate test for this. But perhaps we could then reuse what we run in canary workflow. (see comment).

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+  CloudFormationClient,
+  DeleteStackCommand,
+} from '@aws-sdk/client-cloudformation';
+import { BackendIdentifierConversions } from '@aws-amplify/platform-core';
+import { getClientConfigPath } from '@aws-amplify/client-config';
+import { TestBranch, amplifyAppPool } from '../amplify_app_pool.js';
+import { testConcurrencyLevel } from '../test-e2e/test_concurrency.js';
+import { e2eToolingClientConfig } from '../e2e_tooling_client_config.js';
+
+type PackageManagerExecutable = 'npm' | 'yarn-classic' | 'yarn-modern' | 'pnpm';
```
Comment (sobolk) [inline packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:20]:
```suggestion
type PackageManager = 'npm' | 'yarn-classic' | 'yarn-modern' | 'pnpm';
```

This is not an executable. We derive executable from this enumeration later.

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+  : testConcurrencyLevel;
+
+const packageManagerSetup = async (
+  packageManagerExecutable: PackageManagerExecutable,
+  dir?: string
+) => {
+  const execaOptions = {
+    cwd: dir || os.homedir(),
+    stdio: 'inherit' as const,
+  };
```
Comment (sobolk) [inline packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:33]: Could we refactor the code in such a way that `process.env.PACKAGE_MANAGER_EXECUTABLE` is read only once somewhere?

Options:
1. Do that in `packageManagerSetup` and have `packageManagerSetup` return a type that provides all what test needs for further execution i.e. `type PackageManagerTestConfiguration = { executable: string; maxConcurrency?: number }`
2. Create `TestPackageManagerControllerFactory` and have `TestPackageManagerController` and subclasses (like we did in create amplify) to encapsulate functionalities.

(I think 1. is good enough for tests).

Also. If this stays as `packageManagerSetup` then it should be renamed to `setupPackageManager` and perhaps live in its own file given volume.

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+    ]);
+    await execa(packageManagerExecutable, ['config', 'get', 'registry']);
+  }
+};
+
+void describe('amplify', { concurrency: concurrency }, () => {
+  const { PACKAGE_MANAGER_EXECUTABLE = 'npm' } = process.env;
+  const packageManagerExecutable = PACKAGE_MANAGER_EXECUTABLE.startsWith('yarn')
+    ? 'yarn'
+    : PACKAGE_MANAGER_EXECUTABLE;
```
Comment (sobolk) [inline packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:83]: this part should be in `packageManagerSetup` (see also the other comment)

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+        stdio: 'inherit',
+      });
+    }
+
+    // nuke the npx cache to ensure we are installing packages from the npm proxy
+    if (PACKAGE_MANAGER_EXECUTABLE !== 'yarn-modern') {
+      await packageManagerSetup(
+        PACKAGE_MANAGER_EXECUTABLE as PackageManagerExecutable
+      );
+    }
```
Comment (sobolk) [inline packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:104]: this too should be part of setup package manager.

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+
+  after(async () => {
+    // stop the npm proxy
+    await execa('npm', ['run', 'stop:npm-proxy'], { stdio: 'inherit' });
+  });
+
+  const initialStates = ['empty', 'module', 'commonjs'] as const;
+
+  initialStates.forEach((initialState) => {
+    void describe('installs expected packages and scaffolds expected files', () => {
```
Comment (sobolk) [inline packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:135]: we shouldn't repeat tests from create amplify if we're creating new file.
i.e. we should start with single test that does full create-amplify -> deployment flow here.

which means that we should:
1. remove initialStates, and just start from `empty`
2. remove most of assertions that repeat create amplify test, i.e. it should be similar to https://github.com/aws-amplify/amplify-backend/blob/main/packages/integration-tests/src/test-live-dependency-health-checks/health_checks.test.ts

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+        assert.ok(
+          testBranchDetails.backend?.stackArn?.includes(
+            branchBackendIdentifier.name
+          )
+        );
+      });
+    });
+  });
+
+  void describe('fails fast', () => {
```
Comment (sobolk) [inline packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:398]: we can remove this block too if we're testing in separate file.

**Comment by 0618** [comment]
Continue on https://github.com/aws-amplify/amplify-backend/pull/901

File: packages/integration-tests/src/e2e-flow/e2e_flow.test.ts (hunk 7 above)
Comment (0618) [reply packages/integration-tests/src/e2e-flow/e2e_flow.test.ts:403]: Removed

**Pull request closed without merging**